### PR TITLE
jq, kubectl integration, secrets fixes

### PR DESCRIPTION
**Pull request by longshorej**
This PR adds t hreemain features (also finally figured out why formatting was broken for me, so a fair bit of white-space changes):

### jq

jq: https://stedolan.github.io/jq/

This adds a feature that allows individual resources to be translated with `jq` if `jq` is installed.

It's using stdlib's `system` for now and output redirection until scala native acquires the JDK Process APIs.

This allows the user to arbitrarily add, remove, and modify the output (on an individual resource level).

Note that this is required because we output multiple documents on the stream. Neither `jq` or `yq` support manipulating such output. Thus, I think it's acceptable to have this feature added directly to `rp`.

### kubectl api detection

We invoke `kubectl api-versions` to determine the list of supported versions for the current context. The default API versions are then determined based on this list.

### Secrets

We now populate files with the secret information instead of environment. This allows arbitrary binary data. See:

https://github.com/lightbend/reactive-lib/pull/28
https://github.com/lightbend/sbt-reactive-app

### CODE REVIEW

**Comment by longshorej** [comment]
### `rp version`

```bash
$ rp version
rp (Reactive CLI) 0.4.2-SNAPSHOT
jq support: Available
-> 0
```

### `rp --help`
```bash
$ rp --help | tail -n 8
  --transform-ingress <value>
                           A jq expression that will be applied to Ingress resources. jq must be installed
  --transform-namespaces <value>
                           A jq expression that will be applied to Namespace resources. jq must be installed
  --transform-pod-controllers <value>
                           A jq expression that will be applied to Pod Controller resources. jq must be installed
  --transform-services <value>
                           A jq expression that will be applied to Service resources. jq must be installed
-> 0
```

### Example usage
```bash
 $ rp generate-kubernetes-deployment --generate-ingress "lagom-java-chirper-tooling-example/chirp-impl:1.0.1-SNAPSHOT"    --pod-controller-replicas 2 --transform-ingress '.look.at.this.test = [1, 2, 3]' 
---
{
  "look" : {
    "at" : {
      "this" : {
        "test" : [
          1,
          2,
          3
        ]
      }
    }
  },
  "spec" : {
    "rules" : [
      {
        "http" : {
          "paths" : [
            {
              "path" : "/api/chirps/live",
              "backend" : {
                "serviceName" : "chirpservice",
                "servicePort" : 10000
              }
            },
            {
              "path" : "/api/chirps/history",
              "backend" : {
                "serviceName" : "chirpservice",
                "servicePort" : 10000
              }
            }
          ]
        }
      }
    ]
  },
  "metadata" : {
    "name" : "chirpservice"
  },
  "kind" : "Ingress",
  "apiVersion" : "extensions/v1beta1"
}
-> 0
```

### Attempted usage when `jq` not installed

```bash
$ rp generate-kubernetes-deployment --generate-ingress "lagom-java-chirper-tooling-example/chirp-impl:1.0.1-SNAPSHOT"    --pod-controller-replicas 2 --transform-ingress '.look.at.this.test = [1, 2, 3]' 
[error, com.lightbend.rp.reactivecli.Main$] Resources cannot be translated because jq is not installed
-> 0
``

**Comment by longshorej** [comment]
### Versions test (note: `extensions/v1beta1`)

```bash
$ minikube start --kubernetes-version v1.7.5
Starting local Kubernetes v1.7.5 cluster...
Starting VM...
Getting VM IP address...
Moving files into cluster...
Downloading localkube binary
 138.70 MB / 138.70 MB [============================================] 100.00% 0s
 0 B / 65 B [----------------------------------------------------------]   0.00%
 65 B / 65 B [======================================================] 100.00% 0sSetting up certs...
Connecting to cluster...
Setting up kubeconfig...
Starting cluster components...
Kubectl is now configured to use the cluster.
Loading cached images from config file.
-> 0


$ rp generate-kubernetes-deployment --generate-ingress "lagom-java-chirper-tooling-example/chirp-impl:1.0.1-SNAPSHOT"    --pod-controller-re
plicas 2 
---
{
  "apiVersion" : "extensions/v1beta1",
  "kind" : "Ingress",
  "metadata" : {
    "name" : "chirpservice"
  },
  "spec" : {
    "rules" : [
      {
        "http" : {
          "paths" : [
            {
              "path" : "/api/chirps/live",
              "backend" : {
                "serviceName" : "chirpservice",
                "servicePort" : 10000
              }
            },
            {
              "path" : "/api/chirps/history",
              "backend" : {
                "serviceName" : "chirpservice",
                "servicePort" : 10000
              }
            }
          ]
        }
      }
    ]
  }
}
-> 0
```

**Comment by longshorej** [comment]
### Secrets

Manual test completed successfully with locally published projects.

Binary data secret created with:

`kubectl create secret generic server-p12 --from-file=./testing-self-signed-server.p12`

Sample app setup with:

build.sbt
```
secrets += Secret("server-p12", "testing-self-signed-server.p12")
```

and an Akka HTTP route

```
      path("secret") {
        get(
          onSuccess(SecretReader.get("server-p12", "testing-self-signed-server.p12")) {
            bytes => complete(bytes.fold("nothing")(_.length + "bytes"))
          }
        )
      }
```

Loading the route indicates the secret was loaded correctly.

File: cli/src/main/scala/com/lightbend/rp/reactivecli/runtime/kubernetes/GeneratedKubernetesResource.scala [hunk 1]
```diff
@@ -16,10 +16,22 @@
...
+private[reactivecli] trait GeneratedKubernetesResource extends GeneratedResource[Json] {
+  def json: Json
+  def jqExpression: Option[String]
+
+  def payload: Json = jqExpression.fold(json)(
+    jq(_, json.nospaces)
+      .parse
+      .fold(
+        error => throw new RuntimeException(s"Unable to parse output from jq: $error"),
+        identity))
```
Comment (fsat) [inline cli/src/main/scala/com/lightbend/rp/reactivecli/runtime/kubernetes/GeneratedKubernetesResource.scala]: This is probably a more style related concern, but I'd extract the `jq` call from this trait to the callsite where the resource is being generated.

**Comment by longshorej** [comment]
Thanks @fsat - I've raised #66 to improve our testability. Note that I did add some tests of jq to the existing unit tests, so are unit tests are already testing that functionality (by actually invoking the binary). But yeah, we can do more.

**Pull request merged**